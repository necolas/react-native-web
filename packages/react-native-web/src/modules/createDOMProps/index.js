/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import AccessibilityUtil from '../AccessibilityUtil';
import StyleSheet from '../../exports/StyleSheet';
import styleResolver from '../../exports/StyleSheet/styleResolver';

const emptyObject = {};

const resetStyles = StyleSheet.create({
  ariaButton: {
    cursor: 'pointer'
  },
  button: {
    appearance: 'none',
    backgroundColor: 'transparent',
    color: 'inherit',
    font: 'inherit',
    textAlign: 'inherit'
  },
  heading: {
    font: 'inherit'
  },
  link: {
    backgroundColor: 'transparent',
    color: 'inherit',
    textDecorationLine: 'none'
  },
  list: {
    listStyle: 'none'
  }
});

const pointerEventsStyles = StyleSheet.create({
  auto: {
    pointerEvents: 'auto'
  },
  'box-none': {
    pointerEvents: 'box-none'
  },
  'box-only': {
    pointerEvents: 'box-only'
  },
  none: {
    pointerEvents: 'none'
  }
});

const defaultStyleResolver = style => styleResolver.resolve(style);

const createDOMProps = (component, props, styleResolver) => {
  if (!styleResolver) {
    styleResolver = defaultStyleResolver;
  }

  if (!props) {
    props = emptyObject;
  }

  const {
    accessibilityLabel,
    accessibilityLiveRegion,
    importantForAccessibility,
    placeholderTextColor,
    pointerEvents,
    style: providedStyle,
    testID,
    /* eslint-disable */
    accessible,
    accessibilityComponentType,
    accessibilityRole,
    accessibilityTraits,
    /* eslint-enable */
    ...domProps
  } = props;

  const isDisabled = AccessibilityUtil.isDisabled(props);
  const role = AccessibilityUtil.propsToAriaRole(props);
  const tabIndex = AccessibilityUtil.propsToTabIndex(props);
  const reactNativeStyle = [
    component === 'a' && resetStyles.link,
    component === 'button' && resetStyles.button,
    role === 'heading' && resetStyles.heading,
    component === 'ul' && resetStyles.list,
    role === 'button' && !isDisabled && resetStyles.ariaButton,
    pointerEvents && pointerEventsStyles[pointerEvents],
    providedStyle,
    placeholderTextColor && { placeholderTextColor }
  ];
  const { className, style } = styleResolver(reactNativeStyle);

  if (isDisabled) {
    domProps['aria-disabled'] = true;
  }
  if (importantForAccessibility === 'no-hide-descendants') {
    domProps['aria-hidden'] = true;
  }
  if (accessibilityLabel && accessibilityLabel.constructor === String) {
    domProps['aria-label'] = accessibilityLabel;
  }
  if (accessibilityLiveRegion && accessibilityLiveRegion.constructor === String) {
    domProps['aria-live'] = accessibilityLiveRegion === 'none' ? 'off' : accessibilityLiveRegion;
  }
  if (className && className.constructor === String) {
    domProps.className = domProps.className ? `${domProps.className} ${className}` : className;
  }
  if (component === 'a' && domProps.target === '_blank') {
    domProps.rel = `${domProps.rel || ''} noopener noreferrer`;
  }
  if (role && role.constructor === String && role !== 'label') {
    domProps.role = role;
  }
  if (style) {
    domProps.style = style;
  }
  if (tabIndex) {
    domProps.tabIndex = tabIndex;
  }
  if (testID && testID.constructor === String) {
    domProps['data-testid'] = testID;
  }

  return domProps;
};

export default createDOMProps;
