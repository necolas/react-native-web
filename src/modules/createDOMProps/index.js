/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import AccessibilityUtil from '../AccessibilityUtil';
import StyleSheet from '../../apis/StyleSheet';
import StyleRegistry from '../../apis/StyleSheet/registry';

const emptyObject = {};

const styles = StyleSheet.create({
  buttonReset: {
    appearance: 'none',
    backgroundColor: 'transparent',
    color: 'inherit',
    font: 'inherit',
    textAlign: 'inherit'
  },
  linkReset: {
    backgroundColor: 'transparent',
    color: 'inherit',
    textDecorationLine: 'none'
  },
  listReset: {
    listStyle: 'none'
  }
});

const pointerEventStyles = StyleSheet.create({
  auto: {
    pointerEvents: 'auto'
  },
  'box-none': {
    pointerEvents: 'box-none'
  },
  'box-only': {
    pointerEvents: 'box-only'
  },
  none: {
    pointerEvents: 'none'
  }
});

const resolver = style => StyleRegistry.resolve(style);

const createDOMProps = (rnProps, resolveStyle) => {
  if (!resolveStyle) {
    resolveStyle = resolver;
  }

  const props = rnProps || emptyObject;
  const {
    accessibilityLabel,
    accessibilityLiveRegion,
    accessible,
    importantForAccessibility,
    pointerEvents,
    style: rnStyle,
    testID,
    type,
    /* eslint-disable */
    accessibilityComponentType,
    accessibilityRole,
    accessibilityTraits,
    /* eslint-enable */
    ...domProps
  } = props;

  const role = AccessibilityUtil.propsToAriaRole(props);
  const pointerEventStyle = pointerEvents !== undefined && pointerEventStyles[pointerEvents];
  const reactNativeStyle = [
    (role === 'button' && styles.buttonReset) ||
      (role === 'link' && styles.linkReset) ||
      (role === 'list' && styles.listReset),
    rnStyle,
    pointerEventStyle
  ];
  const { className, style } = resolveStyle(reactNativeStyle) || emptyObject;

  if (accessible === true) {
    domProps.tabIndex = AccessibilityUtil.propsToTabIndex(props);
  }
  if (accessibilityLabel && accessibilityLabel.constructor === String) {
    domProps['aria-label'] = accessibilityLabel;
  }
  if (accessibilityLiveRegion && accessibilityLiveRegion.constructor === String) {
    domProps['aria-live'] = accessibilityLiveRegion === 'none' ? 'off' : accessibilityLiveRegion;
  }
  if (className && className.constructor === String) {
    domProps.className = domProps.className ? `${domProps.className} ${className}` : className;
  }
  if (importantForAccessibility === 'no-hide-descendants') {
    domProps['aria-hidden'] = true;
  }
  if (role && role.constructor === String) {
    domProps.role = role;
    if (role === 'button') {
      domProps.type = 'button';
    } else if (role === 'link' && domProps.target === '_blank') {
      domProps.rel = `${domProps.rel || ''} noopener noreferrer`;
    }
  }
  if (style) {
    domProps.style = style;
  }
  if (testID && testID.constructor === String) {
    domProps['data-testid'] = testID;
  }
  if (type && type.constructor === String) {
    domProps.type = type;
  }

  return domProps;
};

export default createDOMProps;
