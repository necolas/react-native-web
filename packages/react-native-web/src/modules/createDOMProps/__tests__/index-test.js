/* eslint-env jasmine, jest */

import createDOMProps from '..';

const createProps = props => createDOMProps(null, props);

describe('modules/createDOMProps', () => {
  test('prop "accessibilityLabel" becomes "aria-label"', () => {
    const accessibilityLabel = 'accessibilityLabel';
    const props = createProps({ accessibilityLabel });
    expect(props['aria-label']).toEqual(accessibilityLabel);
  });

  test('prop "accessibilityLiveRegion" becomes "aria-live"', () => {
    const accessibilityLiveRegion = 'none';
    const props = createProps({ accessibilityLiveRegion });
    expect(props['aria-live']).toEqual('off');
  });

  describe('prop "accessibilityRole"', () => {
    test('does not become "role" when value is "label"', () => {
      const accessibilityRole = 'label';
      const props = createProps({ accessibilityRole });
      expect(props.role).toBeUndefined();
    });
  });

  test('prop "className" is preserved', () => {
    const className = 'external-class-name';
    const props = createProps({ className });
    expect(props.className).toEqual(className);
  });

  test('prop "importantForAccessibility" becomes "aria-hidden"', () => {
    const props = createProps({ importantForAccessibility: 'no-hide-descendants' });
    expect(props['aria-hidden']).toEqual(true);
  });

  test('prop "testID" becomes "data-testid"', () => {
    const testID = 'Example.testID';
    const props = createProps({ testID });
    expect(props['data-testid']).toEqual(testID);
  });

  test('includes reset styles for "a" elements', () => {
    const props = createDOMProps('a');
    expect(props).toMatchSnapshot();
  });

  test('includes "rel" values for "a" elements (to securely open external links)', () => {
    const props = createDOMProps('a', { target: '_blank' });
    expect(props.rel).toMatchSnapshot();
  });

  test('includes reset styles for "button" elements', () => {
    const props = createDOMProps('button');
    expect(props).toMatchSnapshot();
  });

  test('includes cursor style for "button" role', () => {
    const props = createDOMProps('span', { accessibilityRole: 'button' });
    expect(props).toMatchSnapshot();
  });

  test('includes reset styles for "ul" elements', () => {
    const props = createDOMProps('ul');
    expect(props).toMatchSnapshot();
  });

  test('includes tabIndex when needed', () => {
    const props = createProps({ accessible: true });
    expect(props.tabIndex).toBeDefined();
  });
});
