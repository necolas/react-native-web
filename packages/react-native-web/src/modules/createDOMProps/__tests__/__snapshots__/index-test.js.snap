// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/createDOMProps includes base reset style for browser-styled elements 1`] = `"css-reset-4rbku5"`;

exports[`modules/createDOMProps includes base reset style for browser-styled elements 2`] = `"css-reset-4rbku5"`;

exports[`modules/createDOMProps includes base reset style for browser-styled elements 3`] = `"css-reset-4rbku5"`;

exports[`modules/createDOMProps includes base reset style for browser-styled elements 4`] = `"css-reset-4rbku5"`;

exports[`modules/createDOMProps includes cursor style for pressable roles 1`] = `"css-cursor-18t94o4"`;

exports[`modules/createDOMProps includes cursor style for pressable roles 2`] = `"css-cursor-18t94o4"`;

exports[`modules/createDOMProps prop "accessibilityState" values are "false" 1`] = `
Object {
  "aria-busy": false,
  "aria-checked": false,
  "aria-expanded": false,
  "aria-grabbed": false,
  "aria-invalid": false,
  "aria-modal": false,
  "aria-pressed": false,
  "aria-readonly": false,
  "aria-required": false,
  "aria-selected": false,
}
`;

exports[`modules/createDOMProps prop "accessibilityState" values are "true" 1`] = `
Object {
  "aria-busy": true,
  "aria-checked": true,
  "aria-disabled": true,
  "aria-expanded": true,
  "aria-grabbed": true,
  "aria-hidden": true,
  "aria-invalid": true,
  "aria-modal": true,
  "aria-pressed": true,
  "aria-readonly": true,
  "aria-required": true,
  "aria-selected": true,
  "disabled": true,
  "hidden": true,
}
`;

exports[`modules/createDOMProps prop "accessibilityState" values are "undefined" 1`] = `Object {}`;
